### ISSUE TITLE
gaia and microservices on k8s

### ISSUE THREAD

**Issue by blockspacer** [romange/gaia#20]
I`m interested in C++ usage for microservices running on k8s.

I plan to use Istio as service mesh and Minikube for local dev.

If gaia suits well for microservices, than i want to contribute example app `gaia Shop: Cloud-Native Microservices Demo Application`

Some questions:
* gaia has custom rpc, is it good idea to use it instead of grpc?
* Is it good idea to use gaia for development of something similar to https://github.com/GoogleCloudPlatform/microservices-demo ?

Similar projects:
* Colossus project has example C++ microservice https://github.com/mattpaletta/foobar-social/blob/9bf9e32f6c2853d7681c36bacc2dfab3c6056b77/news_feed/include/main.cpp#L22
* foobar-social - grpc - Colossus - Example Microservices https://github.com/mattpaletta/foobar-social/blob/9bf9e32f6c2853d7681c36bacc2dfab3c6056b77/news_feed/include/main.cpp#L22
* pwave_cloud (grpc++) with istio scrips https://github.com/Toeplitz/pwave_cloud/commit/fc5ae5a786fef80ad4afee7a98cb21db94a47ad2#diff-a055b85bab3694cc2083e4ae4a7d2b50

**Comment by romange** [comment]
gaia provides low-level rpc support with focus on performance and scalability. if you need 1M qps per node then gaia is better suited than grpc. See [asio_fibers](https://github.com/romange/gaia/blob/master/examples/asio_fibers.cc) that demonstrates gaia rpc in action. grpc, on the other hand, has more features and is probably better suited for microservices. Google invested dozens of thousands of eng hours into grpc and it covers many languages. gaia rpc is currently c++ only. The open source version of gaia does not include protobuf bindings at the moment and high-level rpc protocol is not implemented.

It's still possible to use gaia to impelement resful backend however I did not design any high level classes for this.  There is [http_main](https://github.com/romange/gaia/blob/master/util/http/http_main.cc) binary that demonstrates how to register http handlers but again I did not invest much effort into designing a convenient RESTful framework. I am using http mainly for debugging and monitoring purposes. 
There is also https client code that I recently wrote for connecting to GCS API.
 
To summarize, gaia can support everything you need but it's not out of the box. If you need to connect to k8s APIs you need to implement them yourself similarly to how I connected [GCE/GCS](https://github.com/romange/gaia/tree/master/util/gce).
if you need to develop a RESTful backend, you will need to think about what functionality is missing and implement it. Gaia will provide you with the framework to implement asynchronous computations and use multiple cores efficiently. If efficiency is not your priority than golang, python or java is better option :)

**Comment by romange** [comment]
FYI, currently, the only "industrial-grade" open-source descendent of GAIA is GAIA-MR that uses GAIA library underneath to expose high-level MapReduce framework.
